leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java

```diff
@@ -470,33 +470,17 @@ public LeshanServer build() {
 
             // check conflict for private key
             if (privateKey != null) {
-                if (incompleteConfig.getPrivateKey() != null && !incompleteConfig.getPrivateKey().equals(privateKey)) {
-                    throw new IllegalStateException(String.format(
-                            "Configuration conflict between LeshanBuilder and DtlsConnectorConfig.Builder for private key: %s != %s",
-                            privateKey, incompleteConfig.getPrivateKey()));
-                }
 
                 // if in raw key mode and not in X.509 set the raw keys
                 if (certificateChain == null && publicKey != null) {
-                    if (incompleteConfig.getPublicKey() != null && !incompleteConfig.getPublicKey().equals(publicKey)) {
-                        throw new IllegalStateException(String.format(
-                                "Configuration conflict between LeshanBuilder and DtlsConnectorConfig.Builder for public key: %s != %s",
-                                publicKey, incompleteConfig.getPublicKey()));
-                    }
 
-                    dtlsConfigBuilder.setIdentity(privateKey, publicKey);
+                    dtlsConfigBuilder.setCertificateIdentityProvider(new SingleCertificateProvider(privateKey, publicKey));
```

Comment:
I think here we should add a check in case user set a `CertificateIdentityProvider` like we do for  `CertificateVerifier` : 

```java
if (incompleteConfig.getAdvancedCertificateVerifier() != null) {
    if (trustedCertificates != null) {
        throw new IllegalStateException(
                "Configuration conflict between LeshanBuilder and DtlsConnectorConfig.Builder: if a AdvancedCertificateVerifier is set, trustedCertificates must not be set.");
    }
} else {
  // create default Certificate Verifier 
}

Reply:
The code before comes with the assumption, that if there is already something set for the cert identity, that is checked against the leshan config. 

```java
if (incompleteConfig.getPrivateKey() != null && !incompleteConfig.getPrivateKey().equals(privateKey)) {
   throw new IllegalStateException(String.format(
                            "Configuration conflict between LeshanBuilder and DtlsConnectorConfig.Builder for private key: %s != %s",
                            privateKey, incompleteConfig.getPrivateKey()));
}
```

The new interface API doesn't support that check, for Californium itself, there is no need for such a validation.
 
So there are different ways to go:
- provide a leshan specific implementation (or additional extended interface), which does that check (means, it will not work with any `CertificateProvider` but only the extended one).
- remove the check and accept, that different keys are used as configured in leshan
- or just, that's they way I did it. use the leshans certificate and provide them in a  CertificateProvider. that disables the feature to use a own implementation.

Reply:
I proposed a fourth way : 
- if user didn't set  `CertificateProvider` set it automatically as you do, and if user set  a `CertificateProvider`  and a private key/publikey/certificate with Leshan API raise IllegalStateException. (exactly as `CertificateVerifier` example above)

Said differently : user should use the API: Either the simple Leshan one OR the full feature Californium one but not both. 

Is there any problem I missed with this way ? :thinking:

Reply:
From the view point of Californium, OK.
But for leshan, I'm not that sure. For me, the  lines

```
            // check conflict for private key
            if (privateKey != null) {
```

will also disable a "default  verifier to be set". so the docu must then contain a hint, that both are required.

So, I can adapt it, but my feeling is, it will undergo some fine tuning inside leshan when using it.

Reply:
I guess I missed something I need to double check how californium builder works now.

Reply:
OK I think I get it. 

The code should be adapt in a way that : 
- either we set  private key/publikey/certificate  with Leshan API OR CertificateProvider with Scandium API
- either we set truststore with Leshan API OR CertificateVerifier with Scandium API.

Second check should probably not be in the ` if (privateKey != null) {`